github/resource_github_repository.go

```diff
@@ -823,17 +823,19 @@ func flattenSecurityAndAnalysis(securityAndAnalysis *github.SecurityAndAnalysis)
 	return []interface{}{securityAndAnalysisMap}
 }
 
-func expandSecurityAndAnalysis(input []interface{}) *github.Repository {
+func expandSecurityAndAnalysis(input []interface{}, d *schema.ResourceData) *github.Repository {
 	if len(input) == 0 || input[0] == nil {
 		return nil
 	}
 
 	securityAndAnalysis := input[0].(map[string]interface{})
 	update := &github.SecurityAndAnalysis{}
 
-	advancedSecurity := securityAndAnalysis["advanced_security"].([]interface{})[0].(map[string]interface{})
-	update.AdvancedSecurity = &github.AdvancedSecurity{
-		Status: github.String(advancedSecurity["status"].(string)),
+	if d.Get("visibility").(string) != "public" {
```

Comment:
I'm not sure this is the correct approach. We should only apply the `advanced_security` settings if they were explicitly set in the HCL. Right now, setting the value in HCL has no effect (with no warning or anything). 

I recommend using `GetOk` instead of `Get` on "security_and_analysis" and only adding the value if set.

Reply:
> The only time where the `advanced_security` field does not need to be set us for public repositories because it is enabled by default.

That is true as of today. It could change in the future and then suddenly things break in an unexpected way.

> In order to use the `GetOk` approach I would need to set that field as optional which may cause confusion for users in other scenarios.

I'm not sure why you'd need to set it as optional to use `GetOk`? I also think that's correct because it _is_ optional. It's not *required* for public or private repositories:

<img width="314" alt="screenshot-20221213-dUPGmOyy@2x" src="https://user-images.githubusercontent.com/408570/207453094-192f827a-484a-4259-9a3e-2308d70d5405.png">

> I think there is some value in seeing that the `advanced_security` field is set in the HCL even for public repos so practitioners can see at a glance that the repo is set to use advanced security.

This is all rooted in the assumption that "public == advanced security analysis", and I think that's a fundamentally flawed assumption.

Reply:
I think we're talking about two different things here.

> I'm not sure why you'd need to set it as optional to use GetOk? I also think that's correct because it is optional. It's not required for public or private repositories:

The block for `security_and_analysis` is currently optional, if it is not set in the HCL then we don't send the call to configure those settings. However, if you do set the `security_and_analysis` block in your HCL all of the subfields are required `advanced_security`, `secret_scanning`, and `secret_scanning_push_protection`. The issue when setting the `security_and_analysis` block for public repos is it would error out because by default `advanced_security` is enabled and sending that field in the payload would give the error message `"Advanced security is always available for public repos"`. However you can still configure the `secret_scanning`, and `secret_scanning_push_protection` features for a public repo. To do a GetOk check I would need to change the `advanced_security` field to optional which I think would cause some confusion in other scenarios as that field needs to be set to enabled to be able to enable the `secret_scanning`, and `secret_scanning_push_protection` fields.

Reply:
I understand what you're saying, but I don't think this fixes the bug. See https://github.com/integrations/terraform-provider-github/blob/48619d4445cf87921850ef8425ee483defff915d/github/resource_github_repository.go#L635-L644, as an example. If the github_repository resource was already under management and the user deletes the entire `security_and_analysis` section, then Terraform will attempt to update the repo to disable Advanced Security (L638). But that's not valid for public repos.

My point is that you're headed down a path where you must explicitly check for repo visibility for any calls that modify `security_and_analysis`. Not only is this very error-prone, but I think it's the wrong design here.

Reply:
I've pushed a change that will add some logic to check the repo visibility when removing the `security_and_analysis` block and send the appropriate payload. 

Why do you see this as error prone? What approach would you recommend?

Reply:
> Why do you see the check for visibility as error prone?

This comment thread is _literally_ an example of why it's error prone 😄. I think a better pattern is to centralize the logic for compiling the `SecurityAndAnalysis` block into a single function that takes the current state, previous state, and repo visibility into consideration when building the structure. Then use that function everywhere and throw a _ton_ of unit tests at it to make sure it behaves as expected under all possible conditions.